File: tools/bij/internal/workspace/workspace.go
package workspace

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

var defaultPodmanPath = "/opt/podman/bin/podman"

// ResolveWorkspace returns the workspace root containing compose.yaml.
func ResolveWorkspace(override string) (string, error) {
	if override != "" {
		return resolveOverride(override)
	}

	cwd, err := os.Getwd()
	if err != nil {
		return "", fmt.Errorf("resolve workspace: get current directory: %w", err)
	}

	dir, err := filepath.Abs(cwd)
	if err != nil {
		return "", fmt.Errorf("resolve workspace: absolutize %q: %w", cwd, err)
	}

	for {
		if hasComposeFile(dir) {
			return dir, nil
		}

		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}

	return "", fmt.Errorf("resolve workspace: compose.yaml not found from %s upward", cwd)
}

// DetectPodman finds the Podman binary using repository runtime precedence.
func DetectPodman() (string, error) {
	if resolved, ok := resolveExecutable(os.Getenv("CONTAINER_RUNTIME")); ok {
		return resolved, nil
	}

	if isExecutable(defaultPodmanPath) {
		return defaultPodmanPath, nil
	}

	if resolved, err := exec.LookPath("podman"); err == nil && isExecutable(resolved) {
		return resolved, nil
	}

	return "", fmt.Errorf("podman not found; install Podman from https://podman.io")
}

// VerifyCompose confirms that podman compose is available.
func VerifyCompose(podman string) error {
	if podman == "" {
		return fmt.Errorf("verify podman compose: podman path is empty")
	}

	cmd := exec.Command(podman, "compose", "version")
	output, err := cmd.CombinedOutput()
	if err != nil {
		message := strings.TrimSpace(string(output))
		if message == "" {
			message = err.Error()
		}
		return fmt.Errorf("verify podman compose: %s", message)
	}

	return nil
}

func resolveOverride(override string) (string, error) {
	root, err := filepath.Abs(override)
	if err != nil {
		return "", fmt.Errorf("resolve workspace override %q: %w", override, err)
	}

	info, err := os.Stat(root)
	if err != nil {
		return "", fmt.Errorf("resolve workspace override %q: %w", root, err)
	}
	if !info.IsDir() {
		return "", fmt.Errorf("resolve workspace override %q: not a directory", root)
	}
	if !hasComposeFile(root) {
		return "", fmt.Errorf("resolve workspace override %q: compose.yaml not found", root)
	}

	return root, nil
}

func hasComposeFile(dir string) bool {
	info, err := os.Stat(filepath.Join(dir, "compose.yaml"))
	return err == nil && !info.IsDir()
}

func resolveExecutable(candidate string) (string, bool) {
	if candidate == "" {
		return "", false
	}

	if filepath.IsAbs(candidate) || strings.Contains(candidate, string(os.PathSeparator)) {
		abs, err := filepath.Abs(candidate)
		if err != nil {
			return "", false
		}
		if isExecutable(abs) {
			return abs, true
		}
		return "", false
	}

	resolved, err := exec.LookPath(candidate)
	if err != nil {
		return "", false
	}
	if !isExecutable(resolved) {
		return "", false
	}

	return resolved, true
}

func isExecutable(path string) bool {
	info, err := os.Stat(path)
	if err != nil || info.IsDir() {
		return false
	}
	return info.Mode()&0o111 != 0
}

File: crates/bijmantra-runtime/src/workspace.rs
use std::env;
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn resolve_workspace(override_path: Option<PathBuf>) -> anyhow::Result<PathBuf> {
    if let Some(path) = override_path {
        return resolve_override(path);
    }

    let mut dir = env::current_dir()?.canonicalize()?;
    loop {
        if has_compose_file(&dir) {
            return Ok(dir);
        }
        if !dir.pop() {
            break;
        }
    }

    anyhow::bail!("resolve workspace: compose.yaml not found from current directory upward")
}

pub fn detect_podman(configured: Option<&str>) -> anyhow::Result<PathBuf> {
    if let Some(candidate) = configured
        && let Some(path) = resolve_executable(candidate)
    {
        return Ok(path);
    }
    if let Ok(candidate) = env::var("CONTAINER_RUNTIME")
        && let Some(path) = resolve_executable(&candidate)
    {
        return Ok(path);
    }

    #[cfg(unix)]
    {
        let default_path = PathBuf::from("/opt/podman/bin/podman");
        if default_path.is_file() {
            return Ok(default_path);
        }
    }

    for candidate in ["podman", "podman.exe"] {
        if let Some(path) = resolve_executable(candidate) {
            return Ok(path);
        }
    }

    anyhow::bail!("podman not found; install Podman from https://podman.io")
}

pub fn verify_compose(podman: &Path) -> anyhow::Result<()> {
    let output = Command::new(podman).args(["compose", "version"]).output()?;
    if output.status.success() {
        return Ok(());
    }

    let message = String::from_utf8_lossy(&output.stderr).trim().to_string();
    if message.is_empty() {
        anyhow::bail!("podman compose version exited with {}", output.status);
    }
    anyhow::bail!("podman compose version: {message}");
}

fn resolve_override(path: PathBuf) -> anyhow::Result<PathBuf> {
    let root = path.canonicalize()?;
    if !root.is_dir() {
        anyhow::bail!(
            "resolve workspace override {}: not a directory",
            root.display()
        );
    }
    if !has_compose_file(&root) {
        anyhow::bail!(
            "resolve workspace override {}: compose.yaml not found",
            root.display()
        );
    }
    Ok(root)
}

fn has_compose_file(dir: &Path) -> bool {
    dir.join("compose.yaml").is_file()
}

fn resolve_executable(candidate: &str) -> Option<PathBuf> {
    let candidate_path = PathBuf::from(candidate);
    if candidate_path.is_absolute() || candidate.contains(std::path::MAIN_SEPARATOR) {
        return candidate_path
            .canonicalize()
            .ok()
            .filter(|path| path.is_file());
    }

    let paths = env::var_os("PATH")?;
    env::split_paths(&paths)
        .flat_map(|path| executable_candidates(&path, candidate))
        .find(|path| path.is_file())
}

fn executable_candidates(path: &Path, candidate: &str) -> Vec<PathBuf> {
    let direct = path.join(candidate);
    if cfg!(windows) && !candidate.ends_with(".exe") {
        vec![direct, path.join(format!("{candidate}.exe"))]
    } else {
        vec![direct]
    }
}
